### PR TITLE
F# Style Guide (new branch)

### PR DESCRIPTION
**Pull request by cartermp**
The current PR (#4916) gives me a GitHub unicorn every time I click, to the point where I cannot address feedback with the github UI.

cc @dsyme @eiriktsarpalis @bartelink @selketjah @anthonylloyd

### CODE REVIEW

File: docs/fsharp/style-guide/component-design-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,884 @@
+---
+title: F# component design guidelines
+description: Learn the guidelines for writing F# components intended for consumption by other callers.
+author: cartermp
```
Comment (BillWagner) [inline docs/fsharp/style-guide/component-design-guidelines.md]: general comment:

@mairaw has been updating our global and folder metadata. The only tags you need are "title", "description", and "ms.date".  (You are the default author for the F# guide.

You can simplify all these files.

File: docs/fsharp/style-guide/component-design-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,884 @@
...
+    let mutable count = 0
+
+    member this.Next() =
+        count <- count + 1
+        count
+```
+
+#### Use interfaces to group related operations
+
+We recommend you use interface types to represent a set of operations. This is preferred to other options, such as tuples of functions or records of functions.
```
Comment (BillWagner) [inline docs/fsharp/style-guide/component-design-guidelines.md]: Avoid "we", unless this is official Microsoft guidance. Consider just "Use interface types...."

**Review by BillWagner — APPROVED** [review]
This version looks great. I had a couple comments.

Once you look at those, :shipit:

File: docs/fsharp/style-guide/component-design-guidelines.md [hunk 3]
```diff
@@ -187,10 +181,10 @@ type Counter() =
 
 #### Use interfaces to group related operations
 
-We recommend you use interface types to represent a set of operations. This is preferred to other options, such as tuples of functions or records of functions.
+Use interface types to represent a set of operations. This is preferred to other options, such as tuples of functions or records of functions.
 
 ```fsharp
-type Serializer =
+type Serializer<'T> =
     abstract Serialize<'T> : preserveRefEq: bool -> value: 'T -> string
```
Comment (bartelink) [inline docs/fsharp/style-guide/component-design-guidelines.md]: The T's need to go off the methods if they are to be at the type level.

File: docs/fsharp/style-guide/formatting.md [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+let methods2 = System.AppDomain.CurrentDomain.GetAssemblies()
+               |> List.ofArray
+               |> List.map (fun assm -> assm.GetTypes())
+               |> Array.concat
+               |> List.ofArray
+               |> List.map (fun t -> t.GetMethods())
+               |> Array.concat
+
+// OK, but prefer previous
+let methods2 =
```
Comment (dsyme) [inline docs/fsharp/style-guide/formatting.md]: re this:
```
// OK
let methods2 = System.AppDomain.CurrentDomain.GetAssemblies()
               |> List.ofArray
               |> List.map (fun assm -> assm.GetTypes())
               |> Array.concat
               |> List.ofArray
               |> List.map (fun t -> t.GetMethods())
               |> Array.concat
```
This first version is not a good convention in practice and I'd always remove it from the compiler code base. In particular code breaks under rename-refactor.  (I think all the code formatting guidelines should be measured with this in mind).

I haven't read the history of the discussion, so I don't know we're you're at on this.  But I'd recommend putting the second version first, and switching the " OK, but prefer previous"

Comment (bartelink) [reply docs/fsharp/style-guide/formatting.md]: +alot I believe I commented to that effect

My grounds for preferring indenting by one unit on next line are that unindenting when the hard-won indentation is not a multiple of your preferred indentation level often  breaks compilation / varies the meaning in various contexts
(a not unimportant secondary justification is that the further code is indented, the harder it is to review in various contexts.)

Comment (cartermp) [reply docs/fsharp/style-guide/formatting.md]: I'll get a PR rolling

Comment (cartermp) [reply docs/fsharp/style-guide/formatting.md]: I'm going to just remove that first example. I'll want to do a deeper pass in this article as it's mostly just a transcription of the Fantomas doc.

File: docs/fsharp/style-guide/conventions.md [hunk 5]
```diff
@@ -0,0 +1,642 @@
...
+## Use classes to contain values that have side effects
+
+There are many times when initializing a value can have side effects, such as instantiating a context to a database or other remote resource. It is tempting to initialize such things in a module and use it in subsequent functions:
+
+```fsharp
+// This is bad!
+module MyApi =
+    let dep1 = File.ReadAllText "/Users/{your name}/connectionstring.txt"
+    let dep2 = Environment.GetEnvironmentVariable "DEP_2"
+    let dep3 = Random().Next() // Random is not thread-safe
```
[the remark this answers is not in the document]
Comment (dsyme) [reply docs/fsharp/style-guide/conventions.md]: Yes, there's nothing intrinsically thread-unsafe about the code shown, though simple variations on it are not safe, e.g. 
```fsharp
module MyApi =
    let dep3 = Random()
    let getRandom() = dep3.Next() // problems if getRandom is called from multiple threads
```
Thread safety and static data is a tricky topic: many, many F# programs are written single-threaded or use multi-threading in constrained ways (Array.Parallel).  It's also possible to write thread-safe global services (lock or use concurrency-safe data structures and so on). So these things depend a lot on the overall programming context, the ability of the programmer etc.

Perhaps simpler guidance would be `statically initialized data should not include values that are not thread safe if your component will itself use multiple threads`.  

In any case, it's not the presence of side effects in the static initialization code that's the real risk - it's the presence of non-thread-safe objects in the results.

**Pull request merged**